### ISSUE TITLE
How to handle breaking changes to switch cases?

### ISSUE THREAD

**Issue by munificent** [dart-lang/dart_style#1164]
The upcoming patterns support reinterprets all existing switch case constant expressions as patterns. Most constant expressions seen in switches are also valid patterns, so that change is mostly harmless. However, there are [some constant expressions that are not valid patterns][breaking]. This is perfectly fine Dart 2.19 code:

[breaking]: https://github.com/dart-lang/language/blob/master/accepted/future-releases/0546-patterns/feature-specification.md#breaking-existing-switches

```dart
switch (obj) {
  case 1 + 2: print('three'); break;
}
```

But the `1 + 2` is a syntax error in Dart 3.0. This is mostly not a problem for users. Cases like this are rare and users can fix their code when they upgrade their package or library's language version to 3.0.

However, this *is* a real problem for the formatter. The formatter is not currently aware of the language version of any file that it processes. When it parses a file, it just unconditionally tells analyzer to treat the file as being [at the latest language version with all experiments enabled](https://github.com/dart-lang/dart_style/blob/master/lib/src/dart_formatter.dart#L91). This is a little hacky, but has worked fine up until now since language changes were always syntactic additions. The latest version of the language was (syntactically) a superset of all previous ones.

With patterns, that's no longer the case. I'm filing this issue to draw attention to that and start talking about a solution. I can think of two so far:

1.  Make dart_style aware of the language version of each file it parses. This is consistent with other tools which generally know the language version. It future-proofs us from running into this problem again in future versions.

    Users have long asked for the ability to configure formatter options in their pubspec. One reason we haven't done that is that dart_style doesn't even know what a pubspec *is*. Making dart_style read the pubspec to ascertain the language version would also put the infrastructure in place to enable this.

    On the other hand, this adds complexity to how the formatter works. It will have to search surrounding directories looking for pubspecs. It will potentially have to look for language version markers in every file before parsing.

    We'll have to figure out how to handle the ways the formatter can be run where that context isn't available. Right now, you can format strings through the library API. You can also pass code to it directly through stdin. You can format files that aren't part of a pub package. I worry about brittleness.

2.  Extend the parser to parse switch cases that are now invalid. Instead of only parsing cases as patterns when a library is at the latest version, parse to a cover grammar that is the union of the pattern grammar plus the old on-longer-supported expression forms. The latter would report syntax errors, but still be parsed correctly and yield a complete AST. The formatter would then ignore those particular errors and work with the AST.

    This makes the formatter's (and my) life easier. But it adds complexity to the parser and AST. I think most of the AST complexity is already done. `SwitchStatement` contains a list of `SwitchMember`s and there are subclasses of `SwitchMember` for `SwitchCase` (old version with an `Expression` for the case) and `SwitchPatternCase` (new one).

    I don't know how much work it would be in the parser to be able to parse a switch statement containing a mixture of pattern and constant expression cases. It may be that the overall user experience and error recovery is better if we do this. Right now, if I enable patterns and parse:

    ```dart
    switch (c) {
      case 1 + 2: oops;
    }
    ```

    I get:

    ```
    line 2, column 12 of .: Expected to find ':'.
      ╷
    2 │     case 1 + 2: oops;
      │            ^
      ╵
    line 2, column 12 of .: Expected an identifier.
      ╷
    2 │     case 1 + 2: oops;
      │            ^
      ╵
    line 2, column 14 of .: Expected to find ';'.
      ╷
    2 │     case 1 + 2: oops;
      │              ^
      ╵
    line 2, column 15 of .: Expected an identifier.
      ╷
    2 │     case 1 + 2: oops;
      │               ^
      ╵
    line 2, column 15 of .: Unexpected text ':'.
      ╷
    2 │     case 1 + 2: oops;
      │               ^
      ╵
    ```

    It would be better (especially for users migrating their code to Dart 3.0) if the error was more like:

    ```
    line 2, column 10 of .: This constant is expression is not a valid pattern in Dart 3.0.
      ╷
    2 │     case 1 + 2: oops;
      │          ^^^^^
      ╵
    ```

Thoughts @leafpetersen @mit-mit @bwilkerson @jensjoha?

**Comment by bwilkerson** [comment]
It might be overkill, but I would probably use the analyzer's APIs to solve this problem.

1. Create an `AnalysisContextCollection` on the file or directory being formatted.
2. For each file being formatted,
  a. use `contextFor` to get the corresponding `AnalysisContext` from the collection,
  b. use `analysisOptions.contextFeatures` to get a `FeatureSet`, then
  c. pass that feature set into `parseString`.

That should get you the right behavior from the parser.

**Comment by jakemac53** [comment]
> 1. We'll have to figure out how to handle the ways the formatter can be run where that context isn't available. Right now, you can format strings through the library API. You can also pass code to it directly through stdin. You can format files that aren't part of a pub package. I worry about brittleness.

Code generators are a good example of this, many of them throw random strings at the formatter today. However they should generally have the ability to get the language version for the file they are generating. So the library API may need to be updated to accept a language version so tools (like code generators) can provide the proper context. That will be breaking, but 🤷‍♂️.

**Comment by munificent** [comment]
@bwilkerson, I tried your suggestion. (Thanks!) I did some very rough performance tests. The code is [here](https://github.com/dart-lang/dart_style/tree/temp-analysis-context-collection).

I built a snapshot using:

```
dart --snapshot-kind=app-jit --snapshot=format.snapshot tool/command_shell.dart format -o none . --analysis-context
```

Then I ran the formatter on some different directories using both the old code and the new code. Here's the total time in seconds from when the Dart program started running until formatting was done:

```
                  Old   AnalysisContextCollection 
dart_style/       0.28  5.50
dart_style/bin/   0.01  0.03
pkg/analyzer/     5.59  8.26
```

It looks like sometimes it takes a few seconds to create an AnalysisContextCollection. For a tool like `dart format` which is run very frequently, I think that's too big of a hit.

However, it was almost instant when running it on dart_style/bin/. Is that because the subdirectory doesn't have a `.dart_tool/package_config.json` file in it? Does that mean that if I just format a subdirectory of a package, it won't correctly detect the package's language version?

Any details on what's going on under the hood? Or am I holding it wrong?

My concern with having dart_style correctly detect the language version is that doing so might be slow and I fear this confirms it.

Here's the full command-line output:

```
$ time dart format.snapshot format -o none .
Processing directory without using analysis context.
Changed lib/src/io2.dart
Formatted 49 files (1 changed) in 0.28 seconds.
dart format.snapshot format -o none .  0.38s user 0.12s system 124% cpu 0.396 total

$ time dart format.snapshot format -o none . --analysis-context
Took 5243ms to create AnalysisContextCollection.
Changed lib/src/io2.dart
Formatted 49 files (1 changed) in 5.50 seconds.
dart format.snapshot format -o none . --analysis-context  4.71s user 1.12s system 103% cpu 5.618 total

$ time dart format.snapshot format -o none bin
Processing directory without using analysis context.
Formatted 1 file (0 changed) in 0.01 seconds.
dart format.snapshot format -o none bin  0.14s user 0.07s system 136% cpu 0.151 total

$ time dart format.snapshot format -o none bin --analysis-context
Took 20ms to create AnalysisContextCollection.
Formatted 1 file (0 changed) in 0.03 seconds.
dart format.snapshot format -o none bin --analysis-context  0.14s user 0.07s system 141% cpu 0.149 total

$ time dart format.snapshot format -o none ../dart/sdk/pkg/analyzer
Processing directory without using analysis context.
Changed ../dart/sdk/pkg/analyzer/lib/src/dart/analysis/analysis_context_collection.dart
Changed ../dart/sdk/pkg/analyzer/lib/src/dart/analysis/context_builder.dart
Changed ../dart/sdk/pkg/analyzer/lib/src/summary2/macro_application.dart
Changed ../dart/sdk/pkg/analyzer/test/src/fasta/recovery/extra_code_test.dart
Formatted 1425 files (4 changed) in 5.59 seconds.
dart format.snapshot format -o none ../dart/sdk/pkg/analyzer  6.18s user 0.80s system 122% cpu 5.712 total

$ time dart format.snapshot format -o none ../dart/sdk/pkg/analyzer --analysis-context
Took 3882ms to create AnalysisContextCollection.
Changed ../dart/sdk/pkg/analyzer/lib/src/dart/analysis/analysis_context_collection.dart
Changed ../dart/sdk/pkg/analyzer/lib/src/dart/analysis/context_builder.dart
Changed ../dart/sdk/pkg/analyzer/lib/src/summary2/macro_application.dart
Changed ../dart/sdk/pkg/analyzer/test/src/fasta/recovery/extra_code_test.dart
Formatted 1425 files (4 changed) in 8.26 seconds.
dart format.snapshot format -o none ../dart/sdk/pkg/analyzer   9.01s user 1.05s system 119% cpu 8.390 total
```

**Comment by jakemac53** [comment]
The analyzer must be doing some extra work here (maybe doing analysis?) - it does not take seconds just to discover and parse a package_config.json file :). The process is really pretty simple/fast, at most it should have to do two File operations per directory in the path to the file being parsed.

I wonder if this is happening for _each_ file being parsed instead of just once?

**Comment by munificent** [comment]
> I wonder if this is happening for each file being parsed instead of just once?

I hope not. I create a single AnalysisContextCollection with all of the files that will be formatted and then use that same object for all of them.

**Comment by jakemac53** [comment]
Well, part of the issue here is that the formatter can run on entire mono_repos etc. So any file at in any sub directory technically could be a part of a different package (this is even the case for a single package, as flutter packages have example dirs which are themselves packages sometimes).

I am not sure how the analyzer would avoid at least looking for a package config in every subdirectory it finds, and unless its doing some relatively smart caching I could see it doing the full lookup for every file.

**Comment by jakemac53** [comment]
Fwiw you could try manually using the package_config package and doing some smart caching of things as you traverse down the tree of files. A fair bit of extra work though.

Essentially I would start by calling [findPackageConfig](https://pub.dev/documentation/package_config/latest/package_config.package_config/findPackageConfig.html) once for each given directory, and then manually traverse down and check for `.dart_tool/package_config.json` files along the way, overriding the config if one is present.

**Comment by munificent** [comment]
Spent some time talking about this with Dart CLI folks. @natebosch suggested just trying to parse everything at the latest language version. If we get a parse error, downgrade the version to the one before patterns and try again. If that happens to parse without error, then the original error must have been from an invalid switch case, so use the new parse. I've got that implemented and it seems to work well.

**Comment by munificent** [comment]
Fixed in #1177. I'm closing this issue now, but the fix won't be on master until after https://github.com/dart-lang/sdk/issues/51415 is fixed and I've rolled dart_style to a version of analyzer containing that fix.
